src/nuc/envelope.py

```diff
@@ -0,0 +1,135 @@
+            header = json.loads(header)
+        except Exception as ex:
+            raise MalformedNucJwtException("invalid header") from ex
+        if not isinstance(header, dict):
+            raise MalformedNucJwtException(
+                f"invalid JWT header type: {type(header).__name__}"
+            )
+        if header.get("alg") != "ES256K":
+            raise MalformedNucJwtException("invalid JWT algorithm")
+        if len(header) != 1:
+            raise MalformedNucJwtException("unexpected keys in header")
+
+        payload = _base64_decode(raw_payload)
+        token = NucToken.parse(payload)
+
+        signature = _base64_decode(signature)
+
+        return DecodedNucToken(raw_header, raw_payload, signature, token)
+
+    def validate_signature(self):
+        """
+        Validate the signature in this token.
+        """
+
+        public_key = PublicKey(self.token.issuer.public_key, raw=True)
+        payload = f"{self.raw_header}.{self.raw_payload}".encode("utf8")
+        signature = public_key.ecdsa_deserialize_compact(self.signature)
+        if not public_key.ecdsa_verify(payload, signature):
+            raise InvalidSignatureException("signature verification failed")
+
+
+class NucTokenEnvelope:
+    """
+    A NUC token envelope, containing a parsed token along with all its proofs
+    """
+
+    def __init__(self, token: DecodedNucToken, proofs: List[DecodedNucToken]) -> None:
+        self._token = token
+        self._proofs = proofs
+
+    @staticmethod
+    def parse(data: str) -> "NucTokenEnvelope":
+        """
+        Parse a NUC token envelope from its serialized JWT form.
+
+        Note that this only parses the envelope and ensures it is structurally correct. This does not perform
+        any form of signature validation.
+        """
+
+        tokens = data.split("/")
+        if len(tokens) == 0:
+            raise MalformedNucJwtException("no tokens found")
+        token = DecodedNucToken.parse(tokens[0])
+        proofs = [DecodedNucToken.parse(token) for token in tokens[1:]]
+        return NucTokenEnvelope(token, proofs)
+
+    def validate_signatures(self):
+        """
+        Validate the signature in this envelope.
+        """
+
+        for token in [self._token, *self._proofs]:
+            token.validate_signature()
+
+    def token(self) -> NucToken:
+        """
+        Get the token in this envelope.
+        """
+
+        return self._token.token
+
+    def proofs(self) -> List[NucToken]:
+        """
+        Get the proofs in this envelope.
+        """
+
+        return [proof.token for proof in self._proofs]
+
+
+class MalformedNucJwtException(Exception):
+    """
+    An exception thrown when a malformed NUC JWT is parsed.
+    """
+
+
+class InvalidSignatureException(Exception):
+    """
+    An exception thrown when signature verification fails.
+    """
+
+
+def _base64_decode(data: str) -> bytes:
+    padding = 4 - (len(data) % 4)
+    data = data + ("=" * padding)
```

Comment:
Could you explain these two lines?

Reply:
Its because valid b64 strings need to be a multiple of 4 if its not pad it with up to 3 x `=`.

The code could use a comment though :)

Reply:
The issue really is that you're not supposed to use `=` in url safe base64 encoding (because... well, it's not url safe). But python's implementation somehow needs it.